### PR TITLE
WIP: PvP Arena

### PR DESCRIPTION
**Pull request by RafaelVidaurre**
# PvP Arena

This a starting point in regards to the PvP Arena. The idea of this PR is to serve as a starting point for discussion and development of the PvP Arena feature. Keep in mind this draft will probably change drastically over time, especially in regards to storage/data structures as we implement the next functionality.

This draft includes:
* Setup for locally running tests in a reasonable amount of time (in `main` tests take about 4 minutes to start, now they are starting in around 20 seconds). We created an alternate set of migrations to improve the setup time for smart contract tests
* Added a hardhat config **only for testing**, this speeds up tests, adds stack trace for reverts and enables `console.log` inside smart contracts.

Functionalities implemented:
* [x] Misc helpers for the frontend (getting items/characters currently in arena, costs, etc)
* [x] Entering arena with a character
* [x] Finding an opponent
* [x] Setup for easier, faster testing with Stack trace and console.log inside smart contracts
* [x] Performing duel (90%, not yet considering shields to calculate power)
* [x] Withdrawing a character
* [x] Re-rolling opponent
* [x] Rankings (Top 3 players positions and scores tracked)
* [x] Ranking rewards (Top players earn money, seasons, etc)
* [x] Withdrawing duel earnings
* [x] Safely use randoms
* [x] Checks on other contracts (prevent entering raids if character is in PvP arena, etc...)

# Resources
PVP_ Main Arena interactions V4
PvP Arena_ Character states V4

For more information until we consolidate it in one place, also have a look at #636

### CODE REVIEW

**Review by Chriserus — COMMENTED** [review]
Take my Solidity comments with a grain of salt, as I don't have much experience with it, yet :)

File: contracts/PvpArena.sol [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+        skillToken = IERC20(game.skillToken());
+        raids = Raid1(raidContract);
+        randoms = IRandoms(randomsContract);
+
+        // TODO: Tweak these values, they are placeholders
+        wageringFactor = 3;
+        _baseWagerUSD = ABDKMath64x64.divu(500, 100); // $5
+        _tierWagerUSD = ABDKMath64x64.divu(50, 100); // $0.5
+        _rankingsPoolTaxPercent = 15;
+        unattackableSeconds = 2 minutes;
```
[the remark this answers is not in the document]
Comment (RafaelVidaurre) [reply contracts/PvpArena.sol:207]: This is incorrect. Time is still defined in seconds, the fact that it says `minutes` doesn't mean the unit has changed.

`2 minutes = uint256(120)`

File: contracts/PvpArena.sol [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+        isOwnedCharacter(characterID)
+    {
+        require(hasPendingDuel(characterID), "Character is not dueling");
+
+        _assignOpponent(characterID);
+
+        skillToken.transferFrom(
+            msg.sender,
+            address(this),
+            getDuelCost(characterID).div(4)
```
[the remark this answers is not in the document]
Comment (RafaelVidaurre) [reply contracts/PvpArena.sol:294]: Agreed, this should actually be a state variable we can change over time 👌

File: contracts/PvpArena.sol [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+        _updateLastActivityTimestamp(defenderID);
+
+        duelByAttacker[attackerID].isPending = false;
+    }
+
+    /// @dev withdraws a character and its items from the arena.
+    /// if the character is in a battle, a penalty is charged
+    function withdrawFromArena(uint256 characterID)
+        external
+        isOwnedCharacter(characterID)
```
[the remark this answers is not in the document]
Comment (RafaelVidaurre) [reply contracts/PvpArena.sol:503]: Done in the removal function

File: truffle-config.js [hunk 4]
```diff
@@ -57,6 +57,13 @@ module.exports = {
       network_id: process.env.ETH_DEV_RPC_NETWORK_ID || '*',       // Any network (default: none)
       gas: parseInt(process.env.ETH_DEV_RPC_GAS, 10) || 67219750 // required for deploy, otherwise it throws weird require-errors on constructor
     },
+    test: {
+      host: 'localhost',
+      gas: 6721975, // required for deploy, otherwise it throws weird require-errors on constructor
+      port: 9545,
+      network_id: '*', // Any network (default: none)
+      defaultEtherBalance: 1000,
+    },
```
Comment (RafaelVidaurre) [inline truffle-config.js:66]: Remove this

File: truffle-config.js [hunk 5]
```diff
@@ -66,7 +73,7 @@ module.exports = {
         }
       )),
       network_id: 0x61,
-      confirmations: 10,
+      confirmations: 1,
```
Comment (RafaelVidaurre) [inline truffle-config.js:76]: remove this

**Pull request merged**